versioning.cmake

```diff
@@ -0,0 +1,117 @@
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+# -----------------------------------------------------------------------------
+
+
+# This function figures out the version from git and returns Version, Major, Minor, Patch
+#
+#  This function will query git to find the most recent tag by calling 
+#       git describe --tags
+#  If there is a tag on the PR then this will be an exact match and this 
+#  determins the version.  If there have been commits
+#  since the tag it may contain additional info relating to the commit.
+#
+#  Once we have the version, we parse out the Major, Minor, and Patch components
+#  and we write the version into the VERSION file if its different.
+#  The components VERSION, MAJOR, MINOR, and PATCH are returned to
+#  the top level CMakeLists.txt file.  The VERSION file's value is also
+#  integrated into the created version.h file which is part of the 
+#  includes folder in the GitHub binary distribution.
+#
+#  These four components are used to create the GitHub release package during
+#  packaging.   The value in VERSION file will be used by setup.py to set
+#  the version when it creates the wheel file which is the package for PYPI.
+#
+#  If someone should try to build without having cloned the repository
+#  (just downloaded a tag.gz or zip), it will just use the current value 
+#  in the VERSION file.
+
+function(get_versions version major minor patch)
+    # do we have Git installed and are we in a repository?
+	find_package(Git)
+	if(GIT_FOUND)
+		execute_process(COMMAND "${GIT_EXECUTABLE}" log --pretty=format:'%h' -n 1
+	        OUTPUT_VARIABLE GIT_REV
+			OUTPUT_STRIP_TRAILING_WHITESPACE
+	        ERROR_QUIET)
+	endif()
+	# Check whether we got any revision; which isn't
+	# always the case, e.g. when someone downloaded a tar.gz or zip
+	# file from Github instead of a checkout...
+    # or maybe git is not even installed.
+	if ("${GIT_REV}" STREQUAL "")
+	    # use the version already in the VERSION file.
+	    if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/VERSION)
+	       file(READ ${CMAKE_CURRENT_SOURCE_DIR}/VERSION GIT_TAG)
+	    endif()
+	else()
+	    
+	    # see if there as a tag set on this PR
+	    execute_process(
+	        COMMAND "${GIT_EXECUTABLE}" describe --exact-match --tags
+	        OUTPUT_VARIABLE GIT_TAG 
+		    OUTPUT_STRIP_TRAILING_WHITESPACE
+		    ERROR_QUIET)
+    
+	    if ("${GIT_TAG}" STREQUAL "")
+	        execute_process(
+	            COMMAND "${GIT_EXECUTABLE}" describe --tags
```

Comment:
what if the format is: either tag, as above. Or SHA of the commit?

Reply:
SHA of the commit provides no MAJOR, MINOR, or PATCH components.
But we should always have something from `git describe --tags` provided we have git and there is at least one tag in there someplace.

My question is should it be just the MAJOR.MINOR.PATCH which is extracted from `git describe --tags` and then re-combined.   Or should it be just whatever `git describe --tags` gives me which may contain some git info following it?

Reply:
> should it be just whatever git describe --tags gives me which may contain some git info following it?

if you managed to get this cross-platform, it'd be nicer. 
Otherwise we go just for m.m.p